docs: :memo: make resources guide run (#1200)
seedcase-project/seedcase-sprout: docs/guide/resources.qmd

:::

Putting your raw data into a data package makes it easier for yourself
and others to use later on. The steps you'll take to get your data into
the structure used by Sprout are:

1.  Create the properties for the resource, using the original raw data
    as a starting point, and edit as needed.
2.  Create the folders for the new resource within the package and save
    the resource properties in the `datapackage.json` file.
3.  Add your data to the data batches of the resource.
4.  Merge the data batches into a new resource data file.
5.  Re-build the data package's `README.md` file from the updated
    `datapackage.json` file.
6.  If you need to update the properties at a later point, you can use
    `update_resource_properties()` and then write the result to the
    `datapackage.json` file.

```{python setup}
#| include: false

> # This `setup` code chunk creates a package and downloads some raw data.

Reviewer: ```suggestion
# This `setup` code chunk creates a package and downloads some data.
```

Just since "raw" is such a tricky word without a strong meaning and to go in line with what we've been talking about.

from urllib.request import urlretrieve

temp_path = Path(mkdtemp())

> package_path = sp.PackagePath(temp_path / "diabetes-study")

Reviewer: That was my idea too :+1: (or the other option is if we go with the "it's internal/a default argument")

Making a data resource requires that you have data that can be made into
a resource in the first place. Usually, generated or collected data
starts out in a bit of a "raw" shape that needs some working. This work
needs to be done before adding the data as a data package, since Sprout
assumes that the data is already
[tidy](https://design.seedcase-project.org/data/). For this guide, we

> use a raw (but fake) data file that is already tidy and that looks like:

Reviewer: ```suggestion
use a (fake) data file that is already tidy and that looks like:
```

from textwrap import dedent
```

## Extracting resource properties from raw data

You'll start by creating the resource's properties. Before you can store
data in your data package, you need to describe it using properties
(i.e., metadata). The resource's properties are what allow other people
to understand what your data is about and to use it more easily. These
properties also define what it means for data in the resource to be
correct, as all data in the resource must match the properties. While
you can create a resource properties object manually using
`ResourceProperties`, it can be quite intensive and time-consuming if
you, for example, have many columns in your data. To make this process
easier, `extract_resource_properties()` allows you to create an initial
resource properties object by extracting as much information as possible

> from the raw data. Afterwards, you can edit these properties as needed.

Reviewer: ```suggestion
from the raw data. Afterwards, you can edit these properties as needed.
```

Author: Did you want to add anything here?

Reviewer: I have no idea :laughing: Might have been a mistake with a left over comment I forgot to delete.

Now you are ready to use `extract_resource_properties()` to extract the

> resource properties from the raw data. This function tries to infer the

Reviewer: ```suggestion
resource properties from the data. This function tries to infer the
```

> data types from the raw data, but it might not get it right, so make

Reviewer: ```suggestion
data types from the data, but it might not get it right, so make
```

print(sp.check_resource_properties(resource_properties))
```

Time to fill in the missing fields of the resource properties:

```{python}
# TODO: Need to consider/design how editing can be done in an easier, user-friendly way.
# TODO: Add more detail when we know what can and can't be extracted.
resource_properties.title = "Patient Data"

> resource_properties.description = "This data resource contains data about..."

Reviewer: Hmmm, I don't like this way of updating the properties. I guess we'll just have to keep seeing how this looks like and think about some better design for it :thinking:

resource_properties = sp.create_resource_properties(
    path=resource_path,
    properties=resource_properties
)
```

The next step is to write the resource properties to the
`datapackage.json` file. Before they are added, they will be checked to
confirm that they are correctly filled in and that no required fields
are missing. You can use the `path_properties()` helper function to give
you the location of the `datapackage.json` file based on the path to
your package.

```{python}
sp.write_resource_properties(
    path=sp.path_properties(package_path),
    resource_properties=resource_properties,
)
```

::: callout-tip
If you want to see the list of resources available in your data package
via Python code (rather than looking at it directly in the file system),
you can list the contents of the resources folder:

```{python}
# TODO: Make a separate `list_resources()` function for this?
print(list(sp.path_resources(package_path).iterdir()))
```
:::

Let's check the contents of the `datapackage.json` file to see that the
resource properties have been added:

```{python}
pprint(sp.read_properties(sp.path_properties(path=package_path)))
```

> ## Storing a backup of the raw data as a ["batch" file](/docs/design/interface/flows.qmd#save-new-added-or-modified-data-to-batch)

Reviewer: ```suggestion
## Storing a backup of the raw data as a batch file

::: callout-note
See the [flow diagrams](/docs/design/interface/flows.qmd) for a 
simplified flow of steps involved in adding batch files.
:::
```

A header is (often) a link when converted to HTML, and you can't have a link within a link.

you can list the contents of the resources folder:

```{python}
# TODO: Make a separate `list_resources()` function for this?
print(list(sp.path_resources(package_path).iterdir()))
```
:::

Let's check the contents of the `datapackage.json` file to see that the
resource properties have been added:

```{python}
#| eval: false
# TODO: eval when PR #1156 merged
pprint(sp.read_properties(sp.path_properties(path=package_path)))
```

## Storing a backup of the raw data as a "batch" file

When you create a new data resource, or add data to an existing one,
Sprout always stores a backup of each addition or modification as a
"batch" file. All the batch data files are stored in a folder called
`batch/` within the resource's folder. They are merged to create the
final Parquet data file for the resource. Batch files can be useful if
you ever need to re-process the data at a later point, troubleshoot any
issues, update incorrect values, or if you need to compare the stored
batch data to your original raw data (the data before it enters into the
data resource).

As shown above, the data is currently stored in the path called

> `raw_data_path`. Time to store this data in the resource's folder by

Reviewer: This explanation probably would benefit from being moved into the design doc and include a link here to it. Just so we keep a central location for explanations.

print(sp.path_resource_batch_files(1, path=package_path))
```

## Building the Parquet resource data file

Now that you've stored the data as a batch file, you can build the
Parquet file that will be used as the data resource. This Parquet file
is built from all the data in the `batch/` folder. Since there is only
one batch data file stored in the resource's folder, only this one will
be used to build the data resource's Parquet file:

```{python}
#| eval: false
# TODO: eval when function implemented
sp.join_resource_batches(
    data_list=...,

>     resource_properties=resource_properties

Author: Yeah, I didn't really try to make the WIP bits correct :sweat_smile:

## Re-building the README file

One of the last steps to adding a new data resource is to re-build the
`README.md` file of the data package. To allow some flexibility with
what gets added to the README text, this next function will only *build
the text*, but not write it to the file. This allows you to add

> additional information to the README text before writing it to the file.

Reviewer: If you have an idea for how to implement, please make an issue with it :grin:

> additional information to the README text before writing it to the file.

Author: Will have a think!

> additional information to the README text before writing it to the file.

Author: Yeah, my initial thought would be to allow them to give us a path to their template. The default would be the path to our internal template. Or something like that

#| eval: false
# TODO: eval when PR #1156 merged
sp.write_file(
    string=readme_text,
    path=sp.path_readme(path=package_path)
)
```

## Updating resource properties

After having created a resource, you may need to make edits to the
properties. While technically you can do this manually by opening up the
`datapackage.json` file and editing it, we strongly recommend you use
the update functions to do this. These functions help to ensure that the
`datapackage.json` file is still in a correct JSON format and has the
correct fields filled in. You can call the
`update_resource_properties()` function with two arguments: the current
resource properties and a resource properties object representing the
updates you want to make. This latter object will very often be a
partial resource properties, in the sense that it will have only those
fields filled in that you want to update. The function will return an
updated resource properties object. Any field in the object representing
the updates will overwrite the corresponding field in the current
properties object.

```{python}
#| eval: false
# TODO: eval when function implemented
resource_properties = sp.update_resource_properties(

>     current=resource_properties,
>     updates=sp.ResourceProperties(

Reviewer: ```suggestion
    current_properties=resource_properties,
    update_properties=sp.ResourceProperties(
```
